### ISSUE TITLE
[Bug] Trailing slash breaking with websites that don't allow it

### ISSUE THREAD

**Issue by roniemartinez** [spider-rs/spider#118]
Tried crawling a website but it failed because trailing slash is appended even though it is not included in the original URL.

Some websites do not accept this so they return 500. A trailing slash shouldn't be forced into the URL.

<img width="724" alt="image" src="https://github.com/spider-rs/spider/assets/2573537/55ded79a-e8f9-4de5-b36a-492f9427af67">

Cause of error: https://github.com/spider-rs/spider/blob/64a6807d3ba6eb4197a19e143d58e5d896260c3b/spider/src/website.rs#L72-L76

**Comment by j-mendez** [comment]
I think it might make sense for an option to disable trailing slash handling. The trailingSlash is for uniquenesses in order to avoid re-crawling to a page.

**Comment by j-mendez** [comment]
> I think it might make sense for an option to disable trailing slash handling. The trailingSlash is for uniquenesses in order to avoid re-crawling to a page.

We should be able to add the pure url used without the trailing slash for the fix without adding any extra configurations.

**Comment by roniemartinez** [comment]
> The trailingSlash is for uniquenesses in order to avoid re-crawling to a page.

I think URLs with and without trailing slash should be treated as 2 different URLs. Some servers will just do a redirect from 1 to the other and vice versa. At least for other spiders used for SEO treat them as such.

Reference: https://ahrefs.com/blog/trailing-slash/

**Comment by j-mendez** [comment]
> > The trailingSlash is for uniquenesses in order to avoid re-crawling to a page.
> 
> I think URLs with and without trailing slash should be treated as 2 different URLs. Some servers will just do a redirect from 1 to the other and vice versa. At least for other spiders used for SEO treat them as such.
> 
> Reference: https://ahrefs.com/blog/trailing-slash/

Valid, I can put a PR up for this in a couple hours unless you want to.

**Comment by roniemartinez** [comment]
> Valid, I can put a PR up for this in a couple hours unless you want to.

Yes please. Thank you

**Comment by j-mendez** [comment]
I accidentally override the last release with the latest. Fixed in v1.30.0, you may still see the urls logged with the trailing slash but, the network request is performed as is. The [url](https://docs.rs/url/latest/url/) crate auto adds the trailing slash after parsing if the url is the root. There was also an issue with the asset resolution detection that is fixed on the release. Appreciate you giving clear details for the issues, thank you!

**Comment by j-mendez** [comment]
https://github.com/spider-rs/spider/commit/fa66c91e7ce5400d584f7b0a67902750e43649c1

**Comment by roniemartinez** [comment]
@j-mendez I just updated but I am unable to build because of the following 3 errors:

```
   Compiling spider v1.30.1
error[E0308]: mismatched types
   --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:567:40
    |
567 |             if self.is_allowed_default(&link, &blacklist_url) {
    |                     ------------------ ^^^^^ expected `&CompactString`, found `&CaseInsensitiveString`
    |                     |
    |                     arguments to this method are incorrect
    |
    = note: expected reference `&CompactString`
               found reference `&CaseInsensitiveString`
note: method defined here
   --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:144:12
    |
144 |     pub fn is_allowed_default(
    |            ^^^^^^^^^^^^^^^^^^
145 |         &self,
146 |         link: &CompactString,
    |         --------------------

error[E0308]: mismatched types
   --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:568:38
    |
568 |                 let page = Page::new(&link, &client).await;
    |                            --------- ^^^^^ expected `&str`, found `&CaseInsensitiveString`
    |                            |
    |                            arguments to this function are incorrect
    |
    = note: expected reference `&str`
               found reference `&CaseInsensitiveString`
note: associated function defined here
   --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/page.rs:135:18
    |
135 |     pub async fn new(url: &str, client: &Client) -> Self {
    |                  ^^^ ---------

error[E0308]: mismatched types
   --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:577:56
    |
577 |                     .insert(CaseInsensitiveString { 0: link_result });
    |                                                        ^^^^^^^^^^^ expected `CompactString`, found `CaseInsensitiveString`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `spider` due to 3 previous errors
```

**Comment by j-mendez** [comment]
> @j-mendez I just updated but I am unable to build because of the following 3 errors:
> 
> ```
>    Compiling spider v1.30.1
> error[E0308]: mismatched types
>    --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:567:40
>     |
> 567 |             if self.is_allowed_default(&link, &blacklist_url) {
>     |                     ------------------ ^^^^^ expected `&CompactString`, found `&CaseInsensitiveString`
>     |                     |
>     |                     arguments to this method are incorrect
>     |
>     = note: expected reference `&CompactString`
>                found reference `&CaseInsensitiveString`
> note: method defined here
>    --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:144:12
>     |
> 144 |     pub fn is_allowed_default(
>     |            ^^^^^^^^^^^^^^^^^^
> 145 |         &self,
> 146 |         link: &CompactString,
>     |         --------------------
> 
> error[E0308]: mismatched types
>    --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:568:38
>     |
> 568 |                 let page = Page::new(&link, &client).await;
>     |                            --------- ^^^^^ expected `&str`, found `&CaseInsensitiveString`
>     |                            |
>     |                            arguments to this function are incorrect
>     |
>     = note: expected reference `&str`
>                found reference `&CaseInsensitiveString`
> note: associated function defined here
>    --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/page.rs:135:18
>     |
> 135 |     pub async fn new(url: &str, client: &Client) -> Self {
>     |                  ^^^ ---------
> 
> error[E0308]: mismatched types
>    --> /Users/roniemartinez/.cargo/registry/src/github.com-1ecc6299db9ec823/spider-1.30.1/src/website.rs:577:56
>     |
> 577 |                     .insert(CaseInsensitiveString { 0: link_result });
>     |                                                        ^^^^^^^^^^^ expected `CompactString`, found `CaseInsensitiveString`
> 
> For more information about this error, try `rustc --explain E0308`.
> error: could not compile `spider` due to 3 previous errors
> ```

Will take a look shortly after lunch. 👍

**Comment by j-mendez** [comment]
fixed https://github.com/spider-rs/spider/commit/f223a17a7017acd9b7306c7fb2b2e96ab28e1653 released in  `v1.30.2`

**Comment by roniemartinez** [comment]
Thank you @j-mendez It works! 🥇
